src/intervals/interval.jl

```diff
@@ -0,0 +1,173 @@
+
+# Base interval types and utilities
+# ---------------------------------
+bitstype 8 Strand
+
+convert(::Type{Strand}, strand::Uint8) = box(Strand, unbox(Uint8, strand))
+convert(::Type{Uint8}, strand::Strand) = box(Uint8, unbox(Strand, strand))
+
+const STRAND_NA   = convert(Strand, 0b000)
+const STRAND_POS  = convert(Strand, 0b001)
+const STRAND_NEG  = convert(Strand, 0b010)
+const STRAND_BOTH = convert(Strand, 0b011)
+
+function show(io::IO, strand::Strand)
+    if strand == STRAND_NA
+        print(io, "(indeterminate strand)")
+    elseif strand == STRAND_POS
+        print(io, "+")
+    elseif strand == STRAND_NEG
+        print(io, "-")
+    elseif strand == STRAND_BOTH
+        print(io, ".")
+    else
+        print(io, "(undefined strand)")
+    end
+end
+
+
+isless(a::Strand, b::Strand) = convert(Uint8, a) < convert(Uint8, b)
+
+
+# Note, just to be clear: this shadows IntervalTrees.Interval
+@doc """
+A genomic interval specifies interval with some associated metadata.
+""" ->
+immutable Interval{T} <: AbstractInterval{Int64}
+    seqname::String
+    first::Int64
+    last::Int64
+    strand::Strand
+    metadata::T
+end
+
+
+function first(i::Interval)
+    return i.first
+end
+
+
+function last(i::Interval)
+    return i.last
+end
+
+
+function isless{T}(a::Interval{T}, b::Interval{T},
+                   seqname_isless::Function=alphanum_isless)
+    if a.seqname != b.seqname
+        return seqname_isless(a.seqname, b.seqname)
+    elseif a.first != b.first
+        return a.first < b.first
+    elseif a.last != b.last
+        return a.last < b.last
+    elseif a.strand != b.strand
+        return a.strand < b.strand
+    else
+        return a.metadata < b.metadata
+    end
+end
+
+
+@doc """
+Return true if interval `a` entirely precedes `b`.
+""" ->
+function precedes{T}(a::Interval{T}, b::Interval{T},
+                     seqname_isless::Function=alphanum_isless)
+    return seqname_isless(a.seqname, b.seqname) ||
```

Comment:
Rather than silently comparing seqname's I would instead hope that this (and perhaps isless) would instead fail loudly when trying to compare intervals of different sequences?